x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -0,0 +1,142 @@
+        this.clock = clock;
+        this.client = client;
+        this.securityIndex = securityIndex;
+        this.clusterService = clusterService;
+        this.enabled = XPackSettings.API_KEY_SERVICE_ENABLED_SETTING.get(settings);
+    }
+
+    /**
+     * Asynchronously creates a new API key based off of the request and authentication
+     * @param authentication the authentication that this api key should be based off of
+     * @param request the request to create the api key included any permission restrictions
+     * @param listener the listener that will be used to notify of completion
+     */
+    public void createApiKey(Authentication authentication, CreateApiKeyRequest request, ActionListener<CreateApiKeyResponse> listener) {
+        ensureEnabled();
+        if (authentication == null) {
+            listener.onFailure(new IllegalArgumentException("authentication must be provided"));
+        } else {
+            final Instant created = clock.instant();
+            final Instant expiration = getApiKeyExpiration(created, request);
+            final SecureString apiKey = UUIDs.randomBase64UUIDSecureString();
+            final Version version = clusterService.state().nodes().getMinNodeVersion();
+            if (version.before(Version.V_7_0_0_alpha1)) { // TODO(jaymode) change to V6_5_0 on backport!
+                LOGGER.warn("nodes prior to the minimum supported version for api keys {} exist in the cluster; these nodes will not be " +
+                    "able to use api keys", Version.V_7_0_0_alpha1);
+            }
+
+            final char[] keyHash = Hasher.PBKDF2.hash(apiKey);
+            try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+                builder.startObject()
+                    .field("doc_type", "api_key")
+                    .field("creation_time", created.toEpochMilli())
+                    .field("expiration_time", expiration == null ? null : expiration.toEpochMilli());
+
+                byte[] utf8Bytes = null;
+                try {
+                    utf8Bytes = CharArrays.toUtf8Bytes(keyHash);
+                    builder.field("api_key_hash").utf8Value(utf8Bytes, 0, utf8Bytes.length);
+                } finally {
+                    if (utf8Bytes != null) {
+                        Arrays.fill(utf8Bytes, (byte) 0);
+                    }
+                }
+
+                builder.array("role_descriptors", request.getRoleDescriptors())
+                    .field("name", request.getName())
+                    .field("version", version.id)
+                    .startObject("creator")
+                    .field("principal", authentication.getUser().principal())
+                    .field("metadata", authentication.getUser().metadata())
+                    .field("realm", authentication.getLookedUpBy() == null ?
+                        authentication.getAuthenticatedBy().getName() : authentication.getLookedUpBy().getName())
```

Comment:
I guess it depends on how this gets used in the future, but we may need to think about how we deal with realm names, and their potential to change over time.
In particular, out-of-the-box we have "default_file" and "default_native", but once someone adds another realm (e.g. LDAP), they tend to not use "default_native" as the name of their native realm (because we don't tell them to).

Reply:
The reason behind storing the realm is based on the invalidation api and being able to invalidate by realm. It really stinks that we don't have a way to enforce realm name consistency between nodes